src/lib.rs

```diff
@@ -31,3 +31,56 @@ pub use c_shared_slice::*;
 
 mod c_unique_slice;
 pub use c_unique_slice::*;
+
+/// Indicates a type with a layout that is stable across Rust compiler versions.
+///
+/// ## Safety
+/// The type's [Type Layout][type-layout] must fit one of the following:
+/// * [Primitive][prim] layout types of 64-bits or less.
+///   * **Examples:** `i8`, `u32`
+/// * Any zero-sized type (ZST).
+///   * **Examples:** `()`
+/// * [`repr(C)`][repr-c] `struct` or `union` types when all fields are also
+///   `StableLayout`.
+///   * **Examples:** Most `libc` and `winapi` types.
+/// * [`repr(transparent)`][repr-transparent] `struct` or `union` types when the
+///   non-ZST field is also `StableLayout`.
+///   * **Examples:** [`Wrapping<T>`](core::num::Wrapping) where `T: StableLayout`.
+/// * Any other layout that is guaranteed by Rust.
+///   * **Examples:** `&T` and `&mut T` where `T: Sized`.
+///
+/// [type-layout]: https://doc.rust-lang.org/stable/reference/type-layout.html
+/// [prim]:
+/// https://doc.rust-lang.org/stable/reference/type-layout.html#primitive-representations
+/// [repr-c]:
+/// https://doc.rust-lang.org/stable/reference/type-layout.html#the-c-representation
+/// [repr-transparent]:
+/// https://doc.rust-lang.org/stable/reference/type-layout.html#the-transparent-representation
+pub unsafe trait StableLayout { }
+
+unsafe impl StableLayout for u8 { }
+unsafe impl StableLayout for u16 { }
+unsafe impl StableLayout for u32 { }
+unsafe impl StableLayout for u64 { }
+unsafe impl StableLayout for usize { }
+
+unsafe impl StableLayout for i8 { }
+unsafe impl StableLayout for i16 { }
+unsafe impl StableLayout for i32 { }
+unsafe impl StableLayout for i64 { }
+unsafe impl StableLayout for isize { }
+
+unsafe impl StableLayout for bool { }
+unsafe impl StableLayout for char { }
+unsafe impl StableLayout for () { }
+
+unsafe impl<T> StableLayout for core::num::Wrapping<T> where T: StableLayout { }
+
+unsafe impl<T> StableLayout for &T where T: Sized { }
```

Comment:
This, and the other references, are unsound, as if this is for rust<->rust, it's busted as soon as the other end dereferences it.

It should be `where T: Sized + StableLayout`.

Reply:
Yeah I even had it require StableLayout in a previous version, dunno why i forgot to put both as the requirement